_posts/2021-02-22-understanding-non-key-joins-with-quarkus-kafka-streams-extension.adoc

```diff
@@ -0,0 +1,443 @@
+
+----
+http PUT http://localhost:8083/connectors/inventory-connector/config < register-postgres.json
+----
+
+Now run an instance of the `debezium/tooling` image:
+
+----
+docker run --tty --rm \
+    --network kstreams-fk-join-network \
+    debezium/tooling:1.1 \
+----
+
+This image provides several useful tools such as https://github.com/edenhill/kafkacat[kafkacat]. Within the tooling container, run kafkacat to examine the results of the streaming pipeline:
+
+----
+kafkacat -b kafka:9092 -C -o beginning -q \
+    -t customers-with-addresses | jq .
+----
+
+== Running Natively
+Kafka Streams applications can easily be scaled out i.e. the load will be shared amongst multiple instances of the application, each processing just a subset of the partitions of the input topic(s).
+"Running Quarkus applications in native code via GraalVM has a very fast start-up time, the applications use significantly less memory when compiled to native code. This allows you to start as many instances of a Kafka Streams pipeline in parallel in a very memory-efficient way".
+If you want to run this application in `native` mode, set the `QUARKUS_MODE` as `native` and run:
+
+----
+mvn clean package -Pnative
+----
+
+To learn more about running Kafka Streams applications as a native binary, please refer to the https://quarkus.io/guides/kafka-streams#running-natively[reference guide].
+
+== More Insights on the Kafka Streams Extension
+
+The Quarkus extension can also help you address some of the common requirements when building microservices.
+For running your Kafka Streams application in production, you can for instance easily add health checks and metrics for the data pipeline.
+
+https://quarkus.io/guides/microprofile-metrics[Micrometer Metrics] allow applications to gather various metrics and statistics that provide insights into what is happening inside the application.
+Via the MicroProfile Metrics API, these metrics can be exposed via HTTP using JSON format or the OpenMetrics format.
+From there they can be scraped by tools such as https://prometheus.io/[Prometheus] and stored for analysis and visualization.
+"Apart from application-specific metrics, you can utilize built-in metrics exposed by various Quarkus extensions".
+
+To monitor your application using Prometheus registry for an existing Quarkus application you can add the `micrometer-registry-prometheus` extension.
+"This extension loads core micrometer extension as well as additional dependencies required to support Prometheus".
+----
+mvn quarkus:add-extension -Dextensions="micrometer-registry-prometheus"
+----
+
+This extension uses `https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/config/MeterFilter.java[MeterFilter]` instances to customize the metrics emitted by `MeterRegistry` instances.
+The Micrometer extension will detect MeterFilter CDI beans and use them when initializing MeterRegistry instances.
+
+[source,java]
```

Comment:
What is this class needed for? Wouldn't we have Kafka Streams metrics also without it?

Reply:
It can be done only using `MeterRegistry`, but using `MeterFilter` enables additional configurations and transformations, using filters. I thought maybe I can try this. Otherwise we can simply switch to basics.

Reply:
I would suggest to stick to the basics for this one, so to make better the point that you get metrics exported very easily, without the need for much (any?) coding.